django/contrib/admin/helpers.py

```diff
@@ -224,37 +224,49 @@ class InlineAdminFormSet:
     A wrapper around an inline formset for use in the admin system.
     """
     def __init__(self, inline, formset, fieldsets, prepopulated_fields=None,
-                 readonly_fields=None, model_admin=None):
+                 readonly_fields=None, model_admin=None, has_view_permission=True, has_add_permission=True,
+                 has_change_permission=True, has_delete_permission=True):
         self.opts = inline
         self.formset = formset
         self.fieldsets = fieldsets
         self.model_admin = model_admin
         if readonly_fields is None:
             readonly_fields = ()
         self.readonly_fields = readonly_fields
+
+        if has_change_permission:
+            self.viewonly_fields = readonly_fields
+        else:
+            self.viewonly_fields = readonly_fields + flatten_fieldsets(fieldsets)
```

Comment:
We can't just extend it, because they are different depending on if we are in extra rows (with the add permission, where unless a field is explicitly readonly, you can still change it) or if we are in existing rows (where all fields are readonly if we don't have the change permission).

Still that's not very clear in the code isn't very clear and I'll refactor so it's more readable.